Create `IFileSaver.SaveAsync()` and `IFolderPicker.PickAsync()`

VladislavAntonyuk opened the pull request:
<!--
 Hello, and thank you for your interest in contributing to the .NET MAUI Toolkit! 

 Before you submit please check that this work relates to one of the following:
 - Bug fix
    If you haven't yet opened an Issue that reports the bug in detail, provides a reproduction sample, and has been verified + reproduced by a member of the .NET MAUI Toolkit core team, please do that before submitting a Pull Request.
 - Feature/Proposal
    If you haven't yet submitted a Proposal that has been Championed by a .NET MAUI core team member, please instead open a Discussion at https://github.com/communitytoolkit/maui/discussions/new where we can discuss the pros/cons of the feature and its implementation. 
 Any PR submitted that does not fit with the above options will be closed.
 -->

 ### Description of Change ###

Add SaveFileDialog and FolderPicker

 ### Linked Issues ###
 <!-- Provide links to issues here (#35 will link to issue number 35). Ensure that a GitHub issue was created for your bug/proposal and it has been approved/Championed. -->

 - Fixes #634 

 ### PR Checklist ###
 <!--
 Please check all the things you did here and double-check that you got it all, or state why you didn't do something.

 If anything is unclear please do ask :)
 -->
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [x] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/180 <!-- Replace XYZ with your docs PR #. The checkbox will be automatically checked once the docs PR has been merged -->


 ### Additional information ###

 <!-- 
 Please use this to aid the reviewer, this could include stating which platform(s) have been tested
 -->

pictos commented on src/CommunityToolkit.Maui.Core/Extensions/UIViewExtensions.macios.cs:
```diff
@@ -84,4 +84,30 @@ public static class UIViewExtensions
+	/// <summary>
+	/// Get current UI view controller
+	/// </summary>
+	public static UIViewController? GetCurrentUIViewController()
```
why not use the `Microsoft.Maui.ApplicationModel.GetCurrentUIViewController` method?

pictos commented on src/CommunityToolkit.Maui.Core/Extensions/UIViewExtensions.macios.cs:
```diff
@@ -84,4 +84,30 @@ public static class UIViewExtensions
+	/// <summary>
+	/// Get Key Window
+	/// </summary>
+	public static UIWindow? GetKeyWindow()
```
Why not use `Microsoft.Maui.Platform.UIApplicationExtensions.GetKeyWindow` method?

VladislavAntonyuk replied:
it is internal, I didn't know about it.

pictos reviewed (commented):
Before you go too far...

brminnick reviewed (commented):
Thanks Vlad!

Let's avoid using `Essentials` in the namespace to align with the fact that .NET MAUI dropped the `Essentials` namespace. 

If we align our namespaces closely to .NET MAUI, it will make it easier to promote the features from the Toolkit in to .NET MAUI.

I think using `namespace CommunityToolkit.Maui.Storage` would be good, because it  more-closely matches [.NET MAUI's `Microsoft.Maui.Storage.FilePicker` namespace](https://github.com/dotnet/maui/blob/c4d37c881db65f7dacba05e63613ac9b810a7ec1/src/Essentials/src/FilePicker/FilePicker.shared.cs#L10), but I'm totally open to suggestions!

brminnick commented on src/CommunityToolkit.Maui.Core/Interfaces/IFolderPicker.shared.cs:
```diff
@@ -0,0 +1,24 @@
+	/// </summary>
+	/// <param name="cancellationToken"><see cref="CancellationToken"/></param>
+	/// <returns><see cref="Folder"/></returns>
+	Task<Folder?> PickAsync(CancellationToken cancellationToken);
```
Let's return `Task<Folder>` (non-nullable) and throw an Exception if the folder cannot be returned.

This aligns more closely to .NET MAUI's `FilePicker` that does things like `throws throw new PlatformNotSupportedException`

brminnick commented on src/CommunityToolkit.Maui.Core/Interfaces/IFolderPicker.shared.cs:
```diff
@@ -0,0 +1,24 @@
+	/// <param name="initialPath">Initial path</param>
+	/// <param name="cancellationToken"><see cref="CancellationToken"/></param>
+	/// <returns><see cref="Folder"/></returns>
+	Task<Folder?> PickAsync(string initialPath, CancellationToken cancellationToken);
```
Let's return `Task<Folder>` (non-nullable) and throw an Exception if the folder cannot be returned.

This aligns more closely to .NET MAUI's `FilePicker` that does things like `throws throw new PlatformNotSupportedException`

For us, we can leverage `FileNotFoundException` when appropriate.

brminnick commented on src/CommunityToolkit.Maui.Core/Interfaces/ISaveFileDialog.shared.cs:
```diff
@@ -0,0 +1,26 @@
+	/// <param name="stream"><see cref="Stream"/></param>
+	/// <param name="cancellationToken"><see cref="CancellationToken"/></param>
+	/// <returns>True if file is saved, otherwise False</returns>
+	Task<bool> SaveAsync(string initialPath, string fileName, Stream stream, CancellationToken cancellationToken);
```
Instead of returning `Task<bool>`, let's return `Task` and throw an Exception if the file is not saved.

We can create a new Exception like `FileSaveException` or something like that.

This aligns more closely to .NET MAUI's FilePicker that does things like `throws throw new PlatformNotSupportedException`.

brminnick commented on src/CommunityToolkit.Maui.Core/Interfaces/ISaveFileDialog.shared.cs:
```diff
@@ -0,0 +1,26 @@
+	/// <param name="stream"><see cref="Stream"/></param>
+	/// <param name="cancellationToken"><see cref="CancellationToken"/></param>
+	/// <returns>True if file is saved, otherwise False</returns>
+	Task<bool> SaveAsync(string fileName, Stream stream, CancellationToken cancellationToken);
```
Instead of returning `Task<bool>`, let's return `Task` and throw an Exception if the file is not saved.

We can create a new Exception like `FileSaveException` or something like that.

This aligns more closely to .NET MAUI's FilePicker that does things like `throws throw new PlatformNotSupportedException`.

brminnick commented on src/CommunityToolkit.Maui.Core/Primitives/Folder.shared.cs:
```diff
@@ -0,0 +1,17 @@
+	/// <summary>
+	/// Folder path.
+	/// </summary>
+	public string Path { get; init; } = null!;
```
Let's avoid the null-forging operator, `!`.

```suggestion
	public string Path { get; init; } = string.Empty;
```

brminnick commented on src/CommunityToolkit.Maui.Core/Primitives/Folder.shared.cs:
```diff
@@ -0,0 +1,17 @@
+	/// <summary>
+	/// Folder name.
+	/// </summary>
+	public string Name { get; init; } = null!;
```
Let's avoid the null-forging operator, `!`.

```suggestion
	public string Name { get; init; } = string.Empty;
```

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+		}
+		catch (Exception)
+		{
+			// ignored
```
Let's propagate Exceptions up to the developer consuming the library (eg `throw;`) to let them handle it instead of ignoring Exceptions.

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+
+sealed class FileFolderDialog : IDisposable
+{
+	readonly Context? context;
```
`context` will never be null because it is assigned in the constructor

```suggestion
	readonly Context context;
```

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+		this.context = context;
+		sdCardDirectory = Environment.RootDirectory.AbsolutePath;
+
+		try
+		{
+			sdCardDirectory = new File(sdCardDirectory).CanonicalPath;
+		}
+		catch (IOException)
+		{
+		}
```
Why are we ignoring `IOException`? 

I'd prefer to propagate it up to the dev consuming FileFolderDialog so that they can be alerted to it and handle it properly for their app.

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+		dialogBuilder.SetNegativeButton("Cancel", (_, _) => { });
+		dirsDialog = dialogBuilder.Create();
+
+		if (dirsDialog != null)
```
```suggestion
		if (dirsDialog is not null)
```

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+
+		var dialogBuilder = CreateDirectoryChooserDialog(directory, subDirectories, (sender, args) =>
+		{
+			var mDirOld = this.directory;
```
Let's choose a more descriptive variable name than `mDirOld`

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+		});
+
+		dialogBuilder.SetNegativeButton("Cancel", (_, _) => { });
+		dirsDialog = dialogBuilder.Create();
```
Let's choose a more descriptive variable name than `dirsDialog`

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+		titleView1?.Dispose();
+	}
+
+	public async Task<string> GetFileOrDirectoryAsync(string directory, string defaultFileName = "default")
```
Can we pass in a `CancellationToken`?

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+				directories.Add("..");
+			}
+
+			Log.Debug("~~~~", "m_dir=" + directory);
```
Can we remove these debug logs?

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+				throw new FileNotFoundException($"Could not locate {directoryFile}");
+			}
+
+			directoryFile = new File(directory);
```
Can we remove these debug logs?

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+			}
+
+			directoryFile = new File(directory);
+			Log.Debug("~~~~~", $"{nameof(directory)}={directory}");
```
Can we remove these debug logs?

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+			Log.Debug("~~~~~", $"{nameof(directory)}={directory}");
+		}
+
+		Log.Debug("~~~~~", $"{nameof(directory)}={directory}");
```
Can we remove these debug logs?

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+		{
+			LayoutParameters = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MatchParent, ViewGroup.LayoutParams.WrapContent),
+
+			//m_titleView1.setTextAppearance(m_context, android.R.style.TextAppearance_Large);
+			//m_titleView1.setTextColor( m_context.getResources().getColor(android.R.color.black) );
```
Can this comment be deleted?

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+		string defaultFileName = "default")
+	{
+		var dialogBuilder = new AlertDialog.Builder(context);
+		////////////////////////////////////////////////
+		// Create title text showing file select type // 
+		////////////////////////////////////////////////
```
Let's trim this comment to one line. We don't use block comments like this anywhere else
```suggestion
		// Create title text showing file select type
```

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+		return directories;
+	}
+
+	/////////////////////////////////////////////////////////////////////////////////////////////////////////
+	//////                                   START DIALOG DEFINITION                                    //////
+	//////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Let's replace this comment with something more descriptive/helpful, or delete it.

We don't use block comments like this anywhere else

```suggestion
```

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+
+	static bool CreateSubDir(string newDir)
+	{
+		var newDirFile = new File(newDir);
```
Let's avoid abbreviating variable names

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+	TextView? titleView1;
+	string selectedFileName;
+
+	//////////////////////////////////////////////////////
+	// Callback interface for selected directory
+	//////////////////////////////////////////////////////
```
Let's either replace this comment with something more descriptive (like an XML comment) or delete it. 

We don't use block comments like this anywhere else in the code base.

```suggestion
```

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+
+		if (fileSelectionMode is FileSelectionMode.FileSave)
+		{
+			///////////////////////////////
+			// Create New Folder Button  //
+			///////////////////////////////
```
This comment isn't helpful; `new Button` is pretty obvious that we're creating a button
```suggestion
```

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+			titleLayout1.AddView(newDirButton);
+		}
+
+		/////////////////////////////////////////////////////
+		// Create View with folder path and entry text box // 
+		/////////////////////////////////////////////////////
```
```suggestion
		// Create View with folder path and entry text box
```

brminnick commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,435 @@
+			titleLayout.AddView(inputText);
+		}
+
+		//////////////////////////////////////////
+		// Set Views and Finish Dialog builder  //
+		//////////////////////////////////////////
```
```suggestion
		// Set Views and Finish Dialog builder
```

brminnick reviewed (changes requested):
Thanks Vlad!! 

What do you think about propagating more Exceptions up to the developer? This PR is using a lot of try/catch blocks that swallow Exceptions like this:
```
try
{
    // ...
}
catch (IOException)
{
    // nothing in the catch block
}
```

I've left more-specific comments in-line in my review.

JoonghyunCho commented:
@VladislavAntonyuk Thanks for the Tizen files! I'm working on filling those up 😄

JoonghyunCho commented:
Just added Tizen implementation. I tried to keep align with the existing android implementation.

<image src="https://user-images.githubusercontent.com/14328614/203915687-11767b28-0e44-4a22-a60a-3381bbb9d54f.gif" width="250" /> <image src="https://user-images.githubusercontent.com/14328614/203915701-7521e56e-e38a-43e4-9836-11d627965da8.gif" width="250" />

pictos commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,397 @@
+		}
+	}
+
+	ArrayAdapter<string> CreateListAdapter(List<string>? items)
+	{
+		var adapter = new SimpleArrayAdapter(context ?? throw new InvalidOperationException(),
+			Android.Resource.Layout.SelectDialogItem,
+			Android.Resource.Id.Text1, items ?? Array.Empty<string>().ToList());
```
I would like to remove the `.ToList()`, I can see two solutions, one is changing the argument to be `IList` as you can see in the snippet below
```suggestion
	ArrayAdapter<string> CreateListAdapter(IList<string>? items)
	{
		var adapter = new SimpleArrayAdapter(context ?? throw new InvalidOperationException(),
			Android.Resource.Layout.SelectDialogItem,
			Android.Resource.Id.Text1, items ?? Array.Empty<string>());
```

and other solution will be to create a static empty List<string> and use that instead. That way we don't allocate any memory.
```
|         Method |       Mean |     Error |    StdDev | Ratio |   Gen0 | Allocated | Alloc Ratio |
|--------------- |-----------:|----------:|----------:|------:|-------:|----------:|------------:|
|    WithNoCache | 17.2207 ns | 0.0703 ns | 0.0587 ns |  1.00 | 0.0051 |      32 B |        1.00 |
|  WithCacheList |  0.5017 ns | 0.0142 ns | 0.0132 ns |  0.03 |      - |         - |        0.00 |
| WithCacheArray |  0.5133 ns | 0.0098 ns | 0.0092 ns |  0.03 |      - |         - |        0.00 |
```

pictos commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,397 @@
+		return adapter;
+	}
+
+	class SimpleArrayAdapter : ArrayAdapter<string>
```
```suggestion
	sealed class SimpleArrayAdapter : ArrayAdapter<string>
```

pictos commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,397 @@
+			if (v is TextView textView)
+			{
+				// Enable list item (directory) text wrapping
+				if (textView.LayoutParameters != null)
```
```suggestion
				if (textView.LayoutParameters is not null)
```

pictos commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,397 @@
+		public override View GetView(int position, View? convertView, ViewGroup parent)
+		{
+			var v = base.GetView(position, convertView, parent);
+			if (v is TextView textView)
```
can we invert this `if`, for better readability

pictos commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,390 @@
+
+namespace CommunityToolkit.Maui.Storage;
+
+enum FileSelectionMode
```
This looks like to be the same enum from Android, can this be at `shared` lvl?

VladislavAntonyuk replied:
Android implementation is reworked from scratch. Now this enum is only used in this file

brminnick replied:
Oooh! You could maybe use the new `file` scope, introduced in C# 11 https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-11.0/file-local-types

pictos commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,390 @@
+
+		var listItems = GetDirectories(path);
+
+		if (listItems != null)
```
```suggestion
		if (listItems is not null)
```

pictos replied:
can we invert this `if`?

pictos commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,390 @@
+		}
+
+		string[] allFilesAndDirectories = Directory.GetDirectories(directoryPath);
+		allFilesAndDirectories = allFilesAndDirectories.Concat(Directory.GetFiles(directoryPath)).ToArray();
```
Do we need to call `ToArray` here?

pictos commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,390 @@
+		return Directory.Exists(newDirectory);
+	}
+
+	List<string> GetDirectories(string path)
```
can we return `IEnumerable<string>` instead?

pictos commented on src/CommunityToolkit.Maui/Essentials/FolderPicker/FolderPickerImplementation.android.cs:
```diff
@@ -0,0 +1,41 @@
+	static string GetExternalDirectory()
+	{
+		return Platform.CurrentActivity?.GetExternalFilesDir(null)?.ParentFile?.ParentFile?.ParentFile?.ParentFile?.AbsolutePath
+				?? "/storage/emulated/0";
```
this `"/storage/emulated/0"` is always safe? Would be better to ask the user to provide a default path?

VladislavAntonyuk replied:
it is only used in case user wants to use default location (see line 44). We also have a method where user can set his own path

pictos commented on src/CommunityToolkit.Maui/Essentials/FolderPicker/FolderPickerImplementation.macios.cs:
```diff
@@ -0,0 +1,50 @@
+	/// <inheritdoc />
+	public async Task<Folder> PickAsync(string initialPath, CancellationToken cancellationToken)
+	{
+		var documentPickerViewController = new UIDocumentPickerViewController(new[] { UTTypes.Folder })
```
can we cache this array, instead of creating a new one everytime?

pictos commented on src/CommunityToolkit.Maui/Essentials/FolderPicker/FolderPickerImplementation.macios.cs:
```diff
@@ -0,0 +1,50 @@
+
+		currentViewController?.PresentViewController(documentPickerViewController, true, null);
+
+		return await taskCompetedSource.Task;
```
can we just return the `Task` instead of awaiting it here?

pictos commented on src/CommunityToolkit.Maui/Essentials/FolderPicker/FolderPickerImplementation.macios.cs:
```diff
@@ -0,0 +1,50 @@
+{
+	/// <inheritdoc />
+	public async Task<Folder> PickAsync(string initialPath, CancellationToken cancellationToken)
+	{
```
On iOS we don't need to ask for permission?

VladislavAntonyuk replied:
iOS file storage is protected by default. we work in sandbox, so path permission is not needed. Also macOS asks for it automatically when call controller

pictos commented on src/CommunityToolkit.Maui/Essentials/FolderPicker/FolderPickerImplementation.tizen.cs:
```diff
@@ -0,0 +1,46 @@
+				externalDirectory = storage.RootDirectory;
+			}
+		}
+		return externalDirectory ?? "/home/owner/media/";
```
Is "/home/owner/media/" path always valid?

VladislavAntonyuk replied:
I suppose yes. it is the default emulated path, like /storage/emulated/0 on Android

pictos commented on src/CommunityToolkit.Maui/Essentials/SaveFileDialog/SaveFileDialogImplementation.android.cs:
```diff
@@ -0,0 +1,42 @@
+			throw new FileSaveException("Path doesn't exist");
+		}
+
+		await WriteStream(stream, path, cancellationToken);
```
```suggestion
		await WriteStream(stream, path, cancellationToken).ConfigureAwait(false);
```

pictos commented on src/CommunityToolkit.Maui/Essentials/SaveFileDialog/SaveFileDialogImplementation.shared.cs:
```diff
@@ -0,0 +1,21 @@
+{
+	static async Task WriteStream(Stream stream, string filePath, CancellationToken cancellationToken)
+	{
+		await using var fileStream = new FileStream(filePath, FileMode.OpenOrCreate);
```
```suggestion
		await using var fileStream = new FileStream(filePath, FileMode.OpenOrCreate).ConfigureAwait(false);
```

pictos commented on src/CommunityToolkit.Maui/Essentials/SaveFileDialog/SaveFileDialogImplementation.tizen.cs:
```diff
@@ -0,0 +1,46 @@
+			throw new FileSaveException("Path doesn't exist");
+		}
+
+		await WriteStream(stream, path, cancellationToken);
```
```suggestion
		await WriteStream(stream, path, cancellationToken).ConfigureAwait(false);
```

pictos commented on src/CommunityToolkit.Maui/Essentials/SaveFileDialog/SaveFileDialogImplementation.windows.cs:
```diff
@@ -0,0 +1,41 @@
+			throw new FileSaveException("Path doesn't exist");
+		}
+
+		await WriteStream(stream, file.Path, cancellationToken);
```
```suggestion
		await WriteStream(stream, file.Path, cancellationToken).ConfigureAwait(false);
```

pictos reviewed (changes requested):
@VladislavAntonyuk thanks for moving this forward, I've a couple of general questions/suggestions, they are:

1. Why this implementation isn't on `Core` project? I couldn't find any `Microsoft.Maui.Controls` reference on your implementation, may I miss something?
2. I see that on some platforms you've to create the dialog and the strings, but on platforms that you don't need that like Windows, the text on the dialog will be localized and your implementation don't, would be great if you can _fix_ that.
3. I see a lot of events using lambdas, which can cause a lot of memory leaks, since (at least for me) is very hard to say if the objects in the mix have the same lifetime, so to be safe I recommend you use methods or local functions and make sure that you unsubscribe those events.
4. About the `Directory` logic, can we use the System.IO at .net6 level, that way we can have one implementation to rule all platforms, making it easier to maintain.

VladislavAntonyuk commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.android.cs:
```diff
@@ -0,0 +1,397 @@
+
+namespace CommunityToolkit.Maui.Storage;
+
+enum FileSelectionMode
```
It is now used only in Tizen file. we can keep it there

pictos replied:
I prefer to move it to a separated file, it's hard to ask external contributors to follow a pattern if we don't have one. Even if this Tizen only, we should move it to a `primitive` folders or something like that

JoonghyunCho commented:
Let me test on some different Tizen devices to see if it is better for Tizen to switch dialogs to system program like your Android updates, or just to stay as is.

JoonghyunCho commented:
> Let me test on some different Tizen devices to see if it is better for Tizen to switch dialogs to system program like your Android updates, or just to stay as is.

Decided to stay as is to cover as many device types as possible.

pictos commented on src/CommunityToolkit.Maui/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -44,28 +33,36 @@ sealed class FileFolderDialog : Popup<string>, IDisposable
+			}
 		}
+
+		return externalDirectory!;
```
we don't use the `!` operator. here you can do something like
```suggestion
		return externalDirectory ?? throw new InvalidOperationException("Huston we have a problem!");
```

VladislavAntonyuk commented on Directory.Build.props:
```diff
@@ -20,7 +20,7 @@
   <ItemGroup>
     <InternalsVisibleTo Include="CommunityToolkit.Maui.UnitTests" />
   </ItemGroup>
-  <PropertyGroup Condition="'$(TF_BUILD)' == 'true' and $([MSBuild]::IsOSPlatform('windows')) == 'true'">
```
I would suggest removing this condition, otherwise, the developer will have a failed build on CI but a green build locally, which may be confusing.
as we only need to install the script once, I don't think it is an issue.

JoonghyunCho replied:
That is weird. Is the CI build failed with including the condition? I expected Tizen to be included in the CI build and excluded on the local build.

JoonghyunCho replied:
Looks like the build will be ok with the condition in both CI and local.

VladislavAntonyuk replied:
If you miss Tizen file on local env, CI will fail.

JoonghyunCho replied:
Oh, you meant when you add a new features? All right, I'm Ok if other members are not skeptical. 😂

pictos commented on Directory.Build.props:
```diff
@@ -3,8 +3,7 @@
     <PublishReadyToRun>false</PublishReadyToRun>
-
-    <LangVersion>latest</LangVersion>
+    <LangVersion>preview</LangVersion>
```
can this break our users somehow? I would suggest to keep it at latest, and open an proposal for this change

VladislavAntonyuk replied:
LangVersion latest uses c# 10, but we have features from c# 11

pictos replied:
C#11 is already stable

pictos commented on src/CommunityToolkit.Maui.UnitTests/Mocks/FolderPickerImplementationMock.cs:
```diff
@@ -0,0 +1,22 @@
+namespace CommunityToolkit.Maui.UnitTests.Mocks;
+class FolderPickerImplementationMock : IFolderPicker
+{
+	public Task<Folder> PickAsync(string initialPath, CancellationToken cancellationToken)
```
Can we change it to be ValueTask?

pictos commented on src/CommunityToolkit.Maui.UnitTests/Mocks/SaveFileDialogImplementationMock.cs:
```diff
@@ -0,0 +1,19 @@
+
+class SaveFileDialogImplementationMock : ISaveFileDialog
+{
+	public Task<string> SaveAsync(string initialPath, string fileName, Stream stream, CancellationToken cancellationToken)
```
Can we change this to be ValueTask?

VladislavAntonyuk commented:
> @VladislavAntonyuk, friendly reminder about those questions that I did for my first review:
> 
> 1.Why this implementation isn't on Core project? I couldn't find any Microsoft.Maui.Controls reference on your implementation, may I miss something?
2. I see that on some platforms you've to create the dialog and the strings, but on platforms that you don't need that like Windows, the text on the dialog will be localized and your implementation don't, would be great if you can fix that.
3.  I see a lot of events using lambdas, which can cause a lot of memory leaks, since (at least for me) is very hard to say if the objects in the mix have the same lifetime, so to be safe I recommend you use methods or local functions and make sure that you unsubscribe those events. 
4. About the Directory logic, can we use the System.IO at .net6 level, that way we can have one implementation to rule all platforms, making it easier to maintain.

1. Fixed
2. Fixed. Created public methods, that should allow users to set their values. It is used only on Tizen.
3. Fixed
4. Fixed.

brminnick reviewed (changes requested):
Thanks Vlad! 

My review just has a couple suggestions, the biggest refactor is around returning `Task` instead of `ValueTask` for `SaveAsync()` and `PickAsync` since all of the platform-specific implementations are using `async`/`await`.

### Possible Naming Change

I want to get your thoughts on changing the names:
- `FolderPicker` -> `Folder`
- `SaveFileDialog` -> `File`

With these new names, we'd have `Folder.PickAsync()` and `File.SaveAsync()` which still make sense. Plus, it would create the opportunity for additional methods in the future like `Folder.DeleteAsync()` and `File.DeleteAsync()`.

pictos commented on samples/CommunityToolkit.Maui.Sample/Platforms/Android/AndroidManifest.xml:
```diff
@@ -2,4 +2,6 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android">
 	<application android:allowBackup="true" android:icon="@mipmap/appicon" android:roundIcon="@mipmap/appicon_round" android:supportsRtl="true"></application>
 	<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
+	<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
+	<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
```
This isn't a good practice. Because if the app doesn't use this feature and reference our lib this permission will be added to their app, and can be refused by app stores.

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/FolderPicker/FolderPickerImplementation.android.cs:
```diff
@@ -0,0 +1,64 @@
+/// <inheritdoc />
+public class FolderPickerImplementation : IFolderPicker
+{
+	const int requestCodeFolderPicker = 12345;
```
This value is an `id` that we can use to check from where the request came, if you override the `OnActivityResult` method. As you can see [here](https://github.com/xamarin/Essentials/blob/main/Xamarin.Essentials/Platform/Platform.android.cs#L32-L35) we need this value when we want a value back from an Intent or Activity.

a sample of [usage](https://stackoverflow.com/a/14148838)

pictos replied:
Also, as Essential does, I recommend us to set a range for all `requestCode`, something starting from 2000 ~ 3000, and not choosing random numbers

VladislavAntonyuk commented:
> Thanks Vlad! 
> 
> 
> 
> My review just has a couple suggestions, the biggest refactor is around returning `Task` instead of `ValueTask` for `SaveAsync()` and `PickAsync` since all of the platform-specific implementations are using `async`/`await`.
> 
> 
> 
> ### Possible Naming Change
> 
> 
> 
> I want to get your thoughts on changing the names:
> 
> - `FolderPicker` -> `Folder`
> 
> - `SaveFileDialog` -> `File`
> 
> 
> 
> With these new names, we'd have `Folder.PickAsync()` and `File.SaveAsync()` which still make sense. Plus, it would create the opportunity for additional methods in the future like `Folder.DeleteAsync()` and `File.DeleteAsync()`.

I suggest keeping the current names. First it follows existing FilePicker from MAUI Essentials, Second we have a class called Folder. We'll need to move all methods to that class. Folder.PickAsync makes sense in case you do it on defined folder. But the idea is about selecting folder. For that we need a folder picker.

VladislavAntonyuk commented on src/CommunityToolkit.Maui.Core/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,399 @@
+﻿using Microsoft.Maui.Controls.Compatibility.Platform.Tizen;
```
removed

VladislavAntonyuk commented on samples/CommunityToolkit.Maui.Sample/Platforms/Tizen/tizen-manifest.xml:
```diff
@@ -1,15 +1,18 @@
+    <shortcut-list />
+    <privileges>
+        <privilege>http://tizen.org/privilege/internet</privilege>
+        <privilege>http://tizen.org/privilege/mediastorage</privilege>
+        <privilege>http://tizen.org/privilege/externalstorage</privilege>
```
This isn't a good practice. Because if the app doesn't use this feature and reference our lib this permission will be added to their app, and can be refused by app stores.

demooooooo commented:
Hi, I am using PickFolder for selecting android folders. This is all working fine. But when I get the path of the file inside the folder, it reports an error that Access to the path '/storage/emulated/0/xxx' is denied. my code to read the file is as follows
`var pickedFolder = await _folderPicker.PickFolder();
var files = Directory.GetFiles(pickedFolder.Path);
           foreach (var item in files)
            {
            }`

VladislavAntonyuk commented:
@demooooooo Hi, .NET Directory class has a limited number of features that work on all platforms. With Android 33 Google implemented a "Virtual path", so you cannot simply traverse all files and folders. You should use Android classes to work with files and folders.
See example https://github.com/CommunityToolkit/Maui/blob/83e30c9870b7a1daa26f5fe5966dbc76d9f7ad01/src/CommunityToolkit.Maui.Core/Essentials/SaveFileDialog/SaveFileDialogImplementation.android.cs#L74 and https://github.com/VladislavAntonyuk/MauiSamples/blob/main/MauiPaint/Platforms/Android/Services/FileDialog.cs

brminnick commented:
Hey Vlad! After giving it some thought, let's change `ISaveFileDialog` to `IFileSaver` and merge this PR. `IFileSaver` better aligns to our `IFolderPicker` and .NET MAUI's `IFilePicker`.

If/when this feature gets promoted into .NET MAUI, the team would likely move `IFileSaver.SaveAsync()` into their existing interface, `IFilePicker.SaveAsync()`, but we can save that discussion for a future day.

Thanks again! And don't forget to update the `IFileSaver` text in the sample app and the docs 👍

brminnick commented on src/CommunityToolkit.Maui.Core/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,398 @@
+
+namespace CommunityToolkit.Maui.Storage;
+
+sealed class FileFolderDialog : Popup<string>
```
Hey @JoonghyunCho! Could you confirm that `FileFolderDialog.tizen.cs` only contains Tizen UI Code and doesn't contain any .NET MAUI UI code. I.e., It doesn't contain any code  from the `Microsoft.Maui.Controls` namespace like `Microsoft.Maui.Controls.Button`.

I think we're good, but I just want to double-check that we're not accidentally putting .NET MAUI UI code in `CommunityToolit.Maui.Core`: 
https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#where-to-make-your-changes

[![](https://user-images.githubusercontent.com/13558917/145694198-7addbd35-0e5f-4816-b351-759a01ec2672.png)](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#where-to-make-your-changes)

JoonghyunCho replied:
Thanks for the double check! No Maui code is included in the `FileFolderDialog.tizen.cs`. It only contains Tizen UI code (Tizen.NUI).

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,398 @@
+				Margin = new Extents(margin1, margin1, 0, 0),
+				WidthSpecification = LayoutParamPolicies.MatchParent
+			};
+			newFolderButton.Clicked += async (sender, args) =>
```
possible memory leak here. The `content` and this button will be alive for how long? Looking is hard to tell the life-time of this object, so would be good to unsubscribe this event when it's no needed anymore.

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,398 @@
+			{
+				try
+				{
+					var newDirectoryName = await new PromptPopup("New Folder Name", "").Open();
```
Is this some title? If so isn't localized and will not be aligned with other platforms that have the dialog out of the box, like Windows

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,398 @@
+				{
+					new Tizen.Applications.ToastMessage
+					{
+						Message = $"Failed to create new folder, {ex.Message}"
```
This isn't localized

JoonghyunCho replied:
Tizen doesn't have the way to get the localized text from the system that's commonly available on all target device profiles... I think I will think about how I can handle this. Or maybe providing the methods or properties to get the text.

VladislavAntonyuk replied:
I added it for 2 strings: Save and OK.
I haven't found a better solution rather then creating a static method to setup values

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,398 @@
+		var okButton = new Button
+		{
+			Focusable = true,
+			Text = isFileSelectionMode ? "Save" : "OK",
```
This string isn't localized... And there're others that I'll not comment to not spam the review

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,398 @@
+
+		var listItems = GetDirectories(path);
+
+		if (listItems is not null)
```
can we invert this if, to better readability

JoonghyunCho replied:
sure

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/FileFolderDialog.tizen.cs:
```diff
@@ -0,0 +1,398 @@
+
+	IEnumerable<string> GetDirectories(string path)
+	{
+		var directories = new List<string>();
```
instead of creating a collection, can we use the `yield return`?

JoonghyunCho replied:
Yes, I've wanted directories to be sorted, but I don't see it is necessary now.

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/FolderPicker/FolderPickerImplementation.android.cs:
```diff
@@ -0,0 +1,63 @@
+		}
+		
+		const string uriSchemeFolder = "content";
+		if (uri.Scheme != null && uri.Scheme.Equals(uriSchemeFolder, StringComparison.OrdinalIgnoreCase))
```
```suggestion
		if (uri.Scheme is not null && uri.Scheme.Equals(uriSchemeFolder, StringComparison.OrdinalIgnoreCase))
```

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/FolderPicker/FolderPickerImplementation.windows.cs:
```diff
@@ -0,0 +1,42 @@
+			folderPickerOperation.Cancel();
+		}
+
+		cancellationToken.Register(CancelFolderPickerOperation);
```
We need to Dispose of the cancellationTokenRegistration, the return type of `Register` method
```suggestion
		var ctsRegistration = cancellationToken.Register(CancelFolderPickerOperation);
		//...
		ctsRegistration.Dispose();
```

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/SaveFileDialog/SaveFileDialogImplementation.android.cs:
```diff
@@ -0,0 +1,88 @@
+		}
+
+		const string uriSchemeFolder = "content";
+		if (uri.Scheme != null && uri.Scheme.Equals(uriSchemeFolder, StringComparison.OrdinalIgnoreCase))
```
```suggestion
		if (uri.Scheme is not null && uri.Scheme.Equals(uriSchemeFolder, StringComparison.OrdinalIgnoreCase))
```

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/SaveFileDialog/SaveFileDialogImplementation.macios.cs:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	/// <inheritdoc />
+	public void Dispose()
```
who calls the Dispose? If it's called when the service is disposed those events will be subscribed to more than once.

pictos commented on src/CommunityToolkit.Maui.Core/Essentials/SaveFileDialog/SaveFileDialogImplementation.windows.cs:
```diff
@@ -0,0 +1,48 @@
+			filePickerOperation.Cancel();
+		}
+
+		cancellationToken.Register(CancelFilePickerOperation);
```
We need to dispose it

pictos reviewed (changes requested):
After another round of reviews, I have those suggestions

dersia commented:
I just saw the discussion about this issue in the communitytoolkit standup video on YouTube and to me this feels a bit weird. This is about a dialog and to me it does not make sense at all that a dialog is responsible for saving files. The dialog should give the UI and return the chosen path to me and saving/opening/deleting etc should happen in a different `FileManger` or just using the bcl. Same for the folder picker, I understand the point of @VladislavAntonyuk that you pick a folder and therefor picker makes sense, but it is also a dialog and therefor I think it should also be called `FolderPickerDialog`.

But this is just me, maybe I am to puristic about this kind of things 😅

VladislavAntonyuk commented:
@dersia thank you for your feedback. FolderPicker is chosen because .NET MAUI already has a FilePicker.
SaveFileDialog is renamed to FileSaver. When user clicks Save button in dialog, I expect to receive a path to the saved file (with its content - no other actions required). Assume Save operation takes 1 minute. User clicked Save button. Then he tries to make changes to the file. But it is not saved yet. Fix me if I am wrong but The result of the operation should already contain a saved file.

dersia commented:
@VladislavAntonyuk thanks for your reply! 

I totally understand your point, but from a DEV perspective, a dialog is a dialog, UI element that gives me an overlay experience over the app, and I see a lot of value in such an UI that is probably extendable by the DEV. However doing file operations like saving, opening (reading) and so on, should in my opionon be something that is done by non UI tech, because I might want that experience by its own, without presenting the user an UI. So taking the case that you are describing, I would rather expect some sort of `FileManager` or such, that gives me the operations for saving, opening and probably SaveAndOoenAsync (I am not sure what I would need that for because for the saving I must have had the content already, right?, but maybe there are use cases I can't think of right now). The responsibility of UI presentation and the file operations are mixed in this. It would be up to be so show a loading spinner until the file is saved and refetched from the io.
The dialog cloud however provide the loading spinner and give the dev a callback function that is called and the spinner is show as part of the dialog UI until the callback finishes (successfully or not), this could also give the dev the opportunity to display an nice looking error UI in case saving or opening failed.

Think of a downloading and save to file scenario. I would present the user a file save dialog, the user gets to choose where to save that file and I download and save the file. Afterwards I would like to directly open the files content, or just use am image and set the image source from the returned path. But this could also be the case for a temporary file, where I would like to have the `SaveAndOpenAsync` method without an dialog or the user to choose. I see value in both things being separate.

With that said, I prefer `IFileSaver` over `IFileDialog` but honestly from a devs perspective I would rather look for something called dialog if I wanted an UI experience that is an dialog.

I don't this feature to be delayed, but I think it makes much more sense to take the time and rethink this and split dialogs and functionality. 😊

brminnick commented:
> from a DEV perspective, a dialog is a dialog

@dersia To clarify - we're not creating any UI; we're not creating any dialogs. 

This PR creates two platform-specific methods, `IFolderPicker.PickAsync()` and `IFileSaver.SaveAsync()`, that enable a developer to pick a folder / save a file using the native iOS/Android/MacCatalyst/Windows/Tizen user interfaces.

In other words, we are not creating any .NET MAUI UI controls.

This PR is similar to [.NET MAUI's `Microsoft.Maui.Storage.IFilePicker`](https://github.com/dotnet/maui/blob/34a8eb02a6921d9883632063eb3e364735330149/src/Essentials/src/FilePicker/FilePicker.shared.cs#L16-L45) which is a platform-specific method that allows the developer to use the native file system to select a file.

I encourage you to check out the [`FileSaver` sample](https://github.com/CommunityToolkit/Maui/pull/699/files#diff-f2e4634fd60ec3ab3fcbb73c59459fba343979b126b97b515617f2bc700cbe61) and [`FolderPicker` sample](https://github.com/CommunityToolkit/Maui/pull/699/files#diff-563d420036113e36e74358330453c0e238ac8d558a79ea0fca5920c604ff6683) that Vlad has added to `CommunityToolkit.Maui.Sample` in this PR to experience and understand the usage of these methods. I also encourage you to watch our monthly standup which are live-streamed to the .NET Foundation YouTube channel to get further insight into this feature as well as other features we are working on: https://github.com/CommunityToolkit/Maui/wiki/Monthly-Standups#january-2023

dersia commented:
Hey @brminnick, thanks for the clarification.

In this case it all makes much more sense. I understood, that there is some UI involved and tgere for the dialog made sense. But if this is just bringing up the native UI, then this sound much more like a `IFileSaver`. 👍🏿

brminnick reviewed (approved):
Thanks Vlad!!! Approved!

marcojak commented:
Congratulations of the approval.
Do you have an ETA for the new Release on nuget?

brminnick commented:
Thanks! No exact date is scheduled yet. 

We'll likely push out a new release next week pending the progress of https://github.com/CommunityToolkit/Maui/pull/667.

j03ybh commented:
Hey @brminnick ,

I am absolutely super excited for this feature. About one month ago we could expect a release depending on the progress of #667 , and in that item I could not see if it was already done or not. 

Is this feature already released, or do you have yet another confirmation/estimation of when we may expect it?

Many thanks in advance!

jfversluis commented:
@j03ybh these features are available today in v4.0.0! In fact, I just release a video about one of them:
https://youtu.be/Q9T-dRYq3Ps

j03ybh commented:
@jfversluis , thank you a bunch!

j03ybh commented:
@jfversluis, 

I tried to download the v4.0.0 nuget, however, it seems it is only compatible with .net7.0. I want to use it in a .NET MAUI Blazor app, and this can only be created with .NET 6.0. Is there a reason v4.0.0 is not compatible with .NET 6? Is this planned to be supported as well?

VladislavAntonyuk commented:
@j03ybn the last version with .net 6 was 2.0.0
Starting with version 3.0.0 we use only .net 7

Maui Blazor works in .net 7 as well.

j03ybh commented:
@VladislavAntonyuk, 

It is my first time rolling with the Community Kit! I did not know it was only supporting .NET  7 since such a while. 

Thanks for pointing out Maui Blazor is also supported, I think I just need to update my Visual Studio to the latest version to be able to choose the .net 7 template, or otherwise I will have to use CLI. 

Thanks for answering a likely silly question then!
